libavcodec/qsvenc.c

```diff
@@ -823,6 +838,47 @@ static int init_video_param(AVCodecContext *avctx, QSVEncContext *q)
         q->extparam_internal[q->nb_extparam_internal++] = (mfxExtBuffer *)&q->extvsi;
     }
 
+    if (q->dual_gfx) {
+#if QSV_HAVE_HE
+        if(q->param.mfx.CodecId != MFX_CODEC_AVC && q->param.mfx.CodecId != MFX_CODEC_HEVC) {
+            av_log(avctx, AV_LOG_WARNING, "Not supported encoder for dual GFX mode. "
+                                          "Supported: h264_qsv and hevc_qsv \n");
+        }
+        if(q->param.mfx.RateControlMethod != MFX_RATECONTROL_VBR &&
+           q->param.mfx.RateControlMethod != MFX_RATECONTROL_CQP &&
+           q->param.mfx.RateControlMethod != MFX_RATECONTROL_ICQ) {
+            av_log(avctx, AV_LOG_WARNING, "Not supported BRC for dual GFX mode. "
+                                          "Supported: VBR, CQP and ICQ \n");
+        }
+        if(q->param.mfx.LowPower != MFX_CODINGOPTION_ON) {
+            av_log(avctx, AV_LOG_WARNING, "Dual GFX mode supports only low-power encoding mode \n");
+        }
+        if((q->param.mfx.CodecId == MFX_CODEC_AVC  && q->param.mfx.IdrInterval != 0) ||
+           (q->param.mfx.CodecId == MFX_CODEC_HEVC && q->param.mfx.IdrInterval != 1)) {
+            av_log(avctx, AV_LOG_WARNING, "Dual GFX mode requires closed GOP (IdrInterval = 0) \n");
+        }
+        if(q->param.mfx.GopPicSize > 30) {
+            av_log(avctx, AV_LOG_WARNING, "For better performance in dual GFX mode GopPicSize must be <= 30 \n");
+        }
+        if(q->param.AsyncDepth < 30) {
+            av_log(avctx, AV_LOG_WARNING, "For better performance in dual GFX mode AsyncDepth must be >= 30 \n");
```

Comment:
I believe it should be errors not logs, and return from the function

Reply:
These checks are also applied in Adaptive mode, which allows to go into a single GPU fallback if some of the parameters do not meet the conditions. If we will return errors here, then this mode will not work. Do we really need to return errors?

As for "For better performance" conditions, there is no definite answer, which parameters should be specified, since recommendations are different for different command line parameters. I specified 30 here, as we recommend in samples, but with some parameters in ffmpeg, I can achieve good performance with lower values. I.e. <30 can be applicable according to the situation.